From 9097fe53d016f537fd3a9d5ba6e19bcd697eed31 Mon Sep 17 00:00:00 2001
From: Mika Laitio <lamikr@pilppa.org>
Date: Tue, 14 May 2024 11:24:33 -0700
Subject: [PATCH 2/2] ROCM_SDK_DEBUG environment variable to print warning

- print warning if environment variable ROCM_SDK_DEBUG
  is set and some of the loaded SO-files does
  not have GPU device architecture specific CO
  code.

Signed-off-by: Mika Laitio <lamikr@pilppa.org>
---
 hipamd/src/hip_fatbin.cpp | 8 +++++++-
 hipamd/src/hip_fatbin.hpp | 1 +
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/hipamd/src/hip_fatbin.cpp b/hipamd/src/hip_fatbin.cpp
index 5d0d00ac9..9ae2fab68 100644
--- a/hipamd/src/hip_fatbin.cpp
+++ b/hipamd/src/hip_fatbin.cpp
@@ -39,7 +39,10 @@ FatBinaryDeviceInfo::~FatBinaryDeviceInfo() {
 FatBinaryInfo::FatBinaryInfo(const char* fname, const void* image) : fdesc_(amd::Os::FDescInit()),
                              fsize_(0), foffset_(0), image_(image), image_mapped_(false),
                              uri_(std::string()) {
-
+  warning_enabled = false;
+  if (getenv("ROCM_SDK_DEBUG")) {
+	  warning_enabled=true;
+  }
   if (fname != nullptr) {
     fname_ = std::string(fname);
   } else {
@@ -251,12 +254,15 @@ hipError_t FatBinaryInfo::ExtractFatBinaryUsingCOMGR(const std::vector<hip::Devi
       auto dev_it = unique_isa_names.find(device_name);
       // If the size is 0, then COMGR API could not find the CO for this GPU device/ISA
       if (dev_it->second.first == 0) {
+        if (warning_enabled)
+          printf("hip_fatbin.cpp: Could not find the CO for %s from file: %s\n", device_name.c_str(), fname_.c_str());
         LogPrintfError("Cannot find CO in the bundle %s for ISA: %s",
                         fname_.c_str(), device_name.c_str());
         hip_status = hipErrorNoBinaryForGpu;
         ListAllDeviceWithNoCOFromBundle(unique_isa_names);
         break;
       }
+      //printf("hip_fatbin.cpp: Found CO for %s from %s\n", device_name.c_str(), fname_.c_str());
       guarantee(unique_isa_names.cend() != dev_it,
                 "Cannot find the device name in the unique device name");
       fatbin_dev_info_[device->deviceId()]
diff --git a/hipamd/src/hip_fatbin.hpp b/hipamd/src/hip_fatbin.hpp
index f9057a4b5..7fea8b298 100644
--- a/hipamd/src/hip_fatbin.hpp
+++ b/hipamd/src/hip_fatbin.hpp
@@ -97,6 +97,7 @@ private:
   amd::Os::FileDesc fdesc_;  //!< File descriptor
   size_t fsize_;             //!< Total file size
   size_t foffset_;           //!< File Offset where the fat binary is present.
+  bool warning_enabled;
 
   // Even when file is passed image will be mmapped till ~desctructor.
   const void* image_;        //!< Image
-- 
2.45.2

